File: src/Tests/Performance/PerfSieve/PerfSieve.cs
namespace go;

using fmt = fmt_package;
using time = time_package;

partial class main_package {

internal static (nint, nint) sieve(nint n) {
    var composite = new slice<bool>(n);
    nint count = 0;
    nint sum = 0;
    for (nint i = 2; i < n; i++) {
        if (!composite[i]) {
            count++;
            sum += i;
            for (nint j = i * i; j < n; j += i) {
                composite[j] = true;
            }
        }
    }
    return (count, sum);
}

internal static void Main() {
    var start = time.Now().UnixNano();
    nint count = 0;
    nint sum = 0;
    for (nint r = 0; r < 3; r++) {
        var (c, s) = sieve(10000000);
        count += c;
        sum += s;
    }
    var elapsed = time.Now().UnixNano() - start;
    fmt.Println("checksum:", count, sum);
    fmt.Println("elapsed_ns:", elapsed);
}

} // end main_package

File: src/Tests/Performance/PerfSieve/PerfSieve.go
package main

import (
	"fmt"
	"time"
)

// Sieve of Eratosthenes: measures slice allocation, indexing, and tight
// integer loops (bounds-check / slice-header emulation overhead in C#).

func sieve(n int) (int, int) {
	composite := make([]bool, n)
	count := 0
	sum := 0
	for i := 2; i < n; i++ {
		if !composite[i] {
			count++
			sum += i
			for j := i * i; j < n; j += i {
				composite[j] = true
			}
		}
	}
	return count, sum
}

func main() {
	start := time.Now().UnixNano()

	count := 0
	sum := 0
	for r := 0; r < 3; r++ {
		c, s := sieve(10000000)
		count += c
		sum += s
	}

	elapsed := time.Now().UnixNano() - start
	fmt.Println("checksum:", count, sum)
	fmt.Println("elapsed_ns:", elapsed)
}
